libtcmu.c

```diff
@@ -317,7 +321,9 @@ static struct nl_sock *setup_netlink(struct tcmulib_context *ctx)
 	 * Could be a older kernel. Ignore failure and just work in degraded
 	 * mode.
 	 */
-	set_genl_features(sock);
+	ret = set_genl_features(sock);
+	if (ret < 0)
+		goto err_unregister;
```

Comment:
We shouldn't fail here, because if we are running on a older kernel runner will fail to start which is too extreme for this optional feature. If that function fails then we just cannot use newer features.

Maybe to make it more user friendly we could just log a more verbose message. Replace the message in set_genl_features with:

"Could not setup kernel features. An update kernel is needed. Will run in degraded mode."

Reply:
Do you means the older kernel cannot support libnl-genl? or other reason? 
If we can confirm the older kernel  just log a more verbose message. If it's not running on older kernel maybe it's better to goto err_unregister.

Reply:
Thanks for your advice. I made a test, the nl_send_sync will return -10 if the kernel not support the TCMU_CMD_SET_FEATURES. 
The libnl transforms the errno to it's own error code.
```
int nl_sendmsg(struct nl_sock *sk, struct nl_msg *msg, struct msghdr *hdr)
{
	ret = sendmsg(sk->s_fd, hdr, 0);
	if (ret < 0)
		return -nl_syserr2nlerr(errno);
}
```
```
int nl_syserr2nlerr(int error)
{
	switch (error) {
	case EOPNOTSUPP:	return NLE_OPNOTSUPP;
        }
}
```

`#define NLE_OPNOTSUPP		10`

In order to catch the error code, it need to define a error code in tcmu-runner.

Reply:
Just to make sure we are on the same page. In libtcmu.c you are just going to include the nl errno.h header like this right:

```
#include <libnl3/netlink/errno.h>

Then in set_genl_features() you will do something like

ret = nl_send_sync()
if (ret == -NLE_OPNOTSUPP) {
```

One more comment. Depending on how you fix this you might need to also fix up the nla_put_u8 ret handling because that also returns a -NLE, but in some places set_genl_features returns -ENOMEM, and we will want to return the same type of code (either a -NLE or -errno) and not a fix of the two.

Reply:
Thanks again. I've thought about this too. If we should return the same type of code, it's better to init the 'ret' to NLE_NOMEM.